ring/subring.go

```diff
@@ -0,0 +1,332 @@
+
+	// Polynomial nb.Coefficients
+	N int
+
+	// Nthroot used for the NTT
+	NthRoot uint64
+
+	// Modulus
+	Modulus uint64
+
+	// Unique factors of Modulus-1
+	Factors []uint64
+
+	// 2^bit_length(Modulus) - 1
+	Mask uint64
+
+	// Fast reduction constants
+	BRedConstant []uint64 // Barrett Reduction
+	MRedConstant uint64   // Montgomery Reduction
+
+	AllowsNTT bool // Indicates whether NTT can be used with the current ring.
+
+	PrimitiveRoot uint64 // 2N-th primitive root
+
+	RootsForward  []uint64 //powers of the 2N-th primitive root in Montgomery form (in bit-reversed order)
+	RootsBackward []uint64 //powers of the inverse of the 2N-th primitive root in Montgomery form (in bit-reversed order)
+	NInv          uint64   //[N^-1] mod Modulus in Montgomery form
+}
+
+// NewSubRing creates a new SubRing with the standard NTT.
+// NTT constants still need to be generated using .GenNTTConstants(NthRoot uint64).
+func NewSubRing(N int, Modulus uint64) (s *SubRing, err error) {
+	return NewSubRingWithCustomNTT(N, Modulus, NumberTheoreticTransformerStandard{}, 2*N)
+}
+
+// NewSubRingWithCustomNTT creates a new SubRing with degree N and modulus Modulus with user-defined NTT transform and primitive Nth root of unity.
+// Modulus should be equal to 1 modulo the root of unity.
+// N must be a power of two larger than 8. An error is returned with a nil *SubRing in the case of non NTT-enabling parameters.
+func NewSubRingWithCustomNTT(N int, Modulus uint64, ntt NumberTheoreticTransformer, NthRoot int) (s *SubRing, err error) {
+
+	// Checks if N is a power of 2
+	if (N < 16) || (N&(N-1)) != 0 && N != 0 {
+		return nil, fmt.Errorf("invalid degree (must be a power of 2 >= 8)")
+	}
+
+	s = &SubRing{}
+
+	s.N = N
+
+	s.NthRoot = uint64(NthRoot)
+
+	s.NumberTheoreticTransformer = ntt
+
+	s.AllowsNTT = false
+
+	s.Modulus = Modulus
+	s.Mask = (1 << uint64(bits.Len64(Modulus-1))) - 1
+
+	// Computes the fast modular reduction constants for the Ring
+	s.BRedConstant = BRedConstant(Modulus)
+
+	// If qi is not a power of 2, we can compute the MRed (otherwise, it
+	// would return an error as there is no valid Montgomery form mod a power of 2)
+	if (Modulus&(Modulus-1)) != 0 && Modulus != 0 {
+		s.MRedConstant = MRedConstant(Modulus)
+	}
+
+	return
+}
+
+// Type returns the Type of subring which might be either `Standard` or `ConjugateInvariant`.
+func (s *SubRing) Type() Type {
+	switch s.NumberTheoreticTransformer.(type) {
+	case NumberTheoreticTransformerStandard:
+		return Standard
+	case NumberTheoreticTransformerConjugateInvariant:
+		return ConjugateInvariant
+	default:
+		panic("invalid NumberTheoreticTransformer type")
+	}
+}
+
+// GenerateNTTConstants generates the NTT constant for the target SubRing.
+// The fields `PrimitiveRoot` and `Factors` can be set manually to
+// bypasse the search for the primitive root (which requires to
+// factor Modulus-1) and speedup the generation of the constants.
+func (s *SubRing) GenerateNTTConstants() (err error) {
```

Comment:
This function also seems to only be used in the package. Probably should be made private like the `ring.Ring` equivalent method.

Reply:
made private